src/sbvr-api/permissions.ts

```diff
@@ -526,31 +527,64 @@ const generateConstrainedAbstractSql = (
 	// permissions circles.
 	const canAccessTrace: string[] = [resourceName];
 
-	const canAccessFunction: ResourceFunction = function (property: AnyObject) {
+	const resolveBind = (maybeBind: any, extraBinds: ODataBinds) => {
+		if (isBindReference(maybeBind)) {
+			const { bind } = maybeBind;
+			if (typeof bind === 'string' || bind < odata.binds.length) {
+				return odata.binds[bind];
+			}
+			return extraBinds[bind - odata.binds.length];
+		}
+		return maybeBind;
+	};
+	const canAccessFunction: ResourceFunction = function (property) {
 		// remove method property so that we won't loop back here again at this point
-		delete property.method;
+		const { method, ...resolvedProperty } = property;
 
 		if (!this.defaultResource) {
 			throw new Error(`No resource selected in AST.`);
 		}
 
 		const targetResource = this.NavigateResources(
 			this.defaultResource,
-			property.name,
+			resolvedProperty.name,
 		);
 
+		const lambdaId = `${lambdaAlias}+${inc}`;
+		inc = inc + 1;
+
 		const targetResourceName = sqlNameToODataName(targetResource.resource.name);
 
 		if (canAccessTrace.includes(targetResourceName)) {
-			// we don't want to allow permission loops for now, therefore we are
-			// throwing the exception here. If we ever want to allow permission
-			// loops return a false AST statement here (like true eq false), to
-			// not recursivley follow query branches in a deep first search.
-			throw new PermissionError(
-				`Permissions for ${resourceName} form a circle by the following path: ${canAccessTrace.join(
-					' -> ',
-				)} -> ${targetResourceName}`,
-			);
+			const { args } = method[1];
+			const depthArg = resolveBind(args[0], this.extraBindVars);
+			if (depthArg == null) {
+				// To enable circular dependencies a depth must be specified and it was not
+				throw new Error(
+					`You must specify a depth if you want to enable circle 'canAccess()' permissions, currently permissions for ${resourceName} form a circle by the following path: ${canAccessTrace.join(
+						' -> ',
+					)} -> ${targetResourceName}`,
+				);
+			}
+			const [type, depth] = depthArg;
+			if (type !== 'Real' || !Number.isInteger(depth) || depth < 1) {
+				throw new Error('The depth for `canAccess` must be an integer >= 1');
+			}
+			if (
+				// Make sure we have enough traces yet to have exceeded the depth before checking if they match
+				canAccessTrace.length > depth &&
+				canAccessTrace
+					.slice(-depth)
+					// Check if we've reached the specified depth with all entries being the same resource, and if so we need to terminate the recursion by returning a falsey expression
+					.every((rName) => rName === targetResourceName)
```

Comment:
Just curious: what's an example that we need all of them to match?
I was expecting something more like `canAccessTrace[canAccessTrace.length - depth] === targetResourceName`.

Reply:
If you were to use `A?A/canAccess(2)` and had `A -> B -> A` then it wouldn't have gone to that depth in a direct circle. That said we probably still want to block indirect circles altogether so I'll have to handle that case differently